kms/capi/capi.go

```diff
@@ -0,0 +1,754 @@
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	sha1Hash := u.Get(HashArg)
+	keyID := u.Get(KeyIDArg)
+	issuerName := u.Get(IssuerNameArg)
+	serialNumber := u.Get(SerialNumberArg)
+
+	// default to the user store
+	var storeLocation string
+	if storeLocation = u.Get(StoreLocationArg); storeLocation == "" {
+		storeLocation = "user"
+	}
+
+	var certStoreLocation uint32
+	switch storeLocation {
+	case "user":
+		certStoreLocation = certStoreCurrentUser
+	case "machine":
+		certStoreLocation = certStoreLocalMachine
+	default:
+		return nil, fmt.Errorf("invalid cert store location %v", storeLocation)
+	}
+
+	var storeName string
+
+	// default to the 'My' store
+	if storeName = u.Get(StoreNameArg); storeName == "" {
+		storeName = "My"
+	}
+
+	st, err := windows.CertOpenStore(
+		certStoreProvSystem,
+		0,
+		0,
+		certStoreLocation,
+		uintptr(unsafe.Pointer(wide(storeName))))
+	if err != nil {
+		return nil, fmt.Errorf("CertOpenStore for the %v store %v returned: %w", storeLocation, storeName, err)
+	}
+
+	var certHandle *windows.CertContext
+
+	switch {
+	case sha1Hash != "":
+		sha1Hash = strings.TrimPrefix(sha1Hash, "0x") // Support specifying the hash as 0x like with serial
+
+		sha1Bytes, err := hex.DecodeString(sha1Hash)
+		if err != nil {
+			return nil, fmt.Errorf("%s must be in hex format: %w", HashArg, err)
+		}
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_SHA1_HASH,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(sha1Bytes)),
+				data: uintptr(unsafe.Pointer(&sha1Bytes[0])),
+			},
+		}
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", HashArg, keyID)
+		}
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	case keyID != "":
+		keyID = strings.TrimPrefix(keyID, "0x") // Support specifying the hash as 0x like with serial
+
+		keyIDBytes, err := hex.DecodeString(keyID)
+		if err != nil {
+			return nil, fmt.Errorf("%v must be in hex format: %w", KeyIDArg, err)
+		}
+		searchData := CERT_ID_KEYIDORHASH{
+			idChoice: CERT_ID_KEY_IDENTIFIER,
+			KeyIDOrHash: CRYPTOAPI_BLOB{
+				len:  uint32(len(keyIDBytes)),
+				data: uintptr(unsafe.Pointer(&keyIDBytes[0])),
+			},
+		}
+		certHandle, err = findCertificateInStore(st,
+			encodingX509ASN|encodingPKCS7,
+			0,
+			findCertID,
+			uintptr(unsafe.Pointer(&searchData)), nil)
+		if err != nil {
+			return nil, fmt.Errorf("findCertificateInStore failed: %w", err)
+		}
+		if certHandle == nil {
+			return nil, fmt.Errorf("certificate with %v=%s not found", KeyIDArg, keyID)
+		}
+		defer windows.CertFreeCertificateContext(certHandle)
+		return certContextToX509(certHandle)
+	case issuerName != "" && serialNumber != "":
```

Comment:
Ok, but add a TODO about [CERT_ISSUER_SERIAL_NUMBER](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_issuer_serial_number)

Reply:
@maraino I can change the search from using `findIssuerStr` to `findIssuerName` which will require users to be explicit about the issuer name and supply it in X.500 format if thats more acceptable?

That SHOULD mean any future fix will only need to solve the SerialNumber search -- i believe its looking for a format that im not aware of